retrofit을 이용하여 github user data 불러오기

june0122 opened the pull request:
<p align = 'center'>
<img width = '200' src = 'https://user-images.githubusercontent.com/39554623/154939566-40f1b8b4-e549-4f3a-9b10-59d317f9d55c.png'>
</p>

retrofit을 이용하여 github user data를 불러오도록 구현하였으며 추가적으로 리사이클러뷰 아이템 더 불러오기(페이징)를 구현 예정입니다.

june0122 committed:
chore: OkHttp Logging Interceptor 라이브러리 추가

june0122 committed:
refactor: Adapter의 아이템 리스트에 들어있던 기본 데이터들 삭제

june0122 committed:
feat: Github에서 데이터를 받아올 HTTP 메서드들을 정의한 인터페이스 정의

june0122 committed:
feat: 응답(response) JSON 데이터를 담을 수 있는 데이터 클래스 추가

june0122 committed:
feat: retrofit으로 Github 사용자 데이터를 가져오기

GithubService 인터페이스 구현체를 통해 Call 객체를 만든 후 `enqueue()` 함수를 실행하여
서버에 API 요청을 보낸다.

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/model/data/Data.kt:
```diff
@@ -1,11 +1,75 @@
+    @field:SerializedName("gravatar_id") val gravatarId: String,
+    @field:SerializedName("html_url") val htmlUrl: String,
+    val id: Int,
+    val login: String,
```
코드 패턴을 위해서 _ 없는 값도 동일하게 해주시는게 좋습니다.

june0122 replied:
```kotlin
data class User(
    @field:SerializedName("avatar_url") val avatarUrl: String,
    @field:SerializedName("events_url") val eventsUrl: String,
    @field:SerializedName("followers_url") val followersUrl: String,
    @field:SerializedName("following_url") val followingUrl: String,
    @field:SerializedName("gists_url") val gistsUrl: String,
    @field:SerializedName("gravatar_id") val gravatarId: String,
    @field:SerializedName("html_url") val htmlUrl: String,
    @field:SerializedName("id") val id: Int,
    @field:SerializedName("login") val login: String,
    @field:SerializedName("node_id") val nodeId: String,
    @field:SerializedName("organizations_url") val organizationsUrl: String,
    @field:SerializedName("received_events_url") val receivedEventsUrl: String,
    @field:SerializedName("repos_url") val reposUrl: String,
    @field:SerializedName("score") val score: Double,
    @field:SerializedName("site_admin") val siteAdmin: Boolean,
    @field:SerializedName("starred_url") val starredUrl: String,
    @field:SerializedName("subscriptions_url") val subscriptionsUrl: String,
    @field:SerializedName("type") val type: String,
    @field:SerializedName("url") val url: String
)
```
가독성을 위해 줄바꿈을 한다던가 더 수정할 부분은 없을까요?

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/fragment/PlantListFragment.kt:
```diff
@@ -73,4 +82,35 @@ class PlantListFragment : Fragment() {
+            override fun onResponse(call: Call<Users>, response: Response<Users>) {
+                if (response.isSuccessful) {
+                    response.body()?.let { users -> updateUserList(users) }
+                }
```
onResponse의 응답값은 UI에서 처리될까요? 다른 스케쥴러에서 동작할까요? `Thread.currentThread()`을 가지고 직접 확인해보세요.

june0122 replied:
Call 객체에서 HTTP request를 보내는 두 가지 방식인 `execute()`와 `enqueue(retrofit2.Callback)` 중 비동기적으로 백그라운드 스레드에서 요청을 보내고 응답이 오면 등록한 콜백 함수를 실행시키는 메서드인 enqueue를 사용하였기 때문에 다른 스케쥴러에서 동작할 것이라고 생각합니다.

페이징 구현하면서 `Thread.currentThread()`로도 직접 확인해보겠습니다.

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/fragment/PlantListFragment.kt:
```diff
@@ -73,4 +82,35 @@ class PlantListFragment : Fragment() {
+                )
+            )
+        }
+        plantListAdapter.notifyItemRangeInserted(0, users.items.size)
```
notifyItemChange()와 notifyItemRangeInserted를 사용한것의 차이가 어떤걸까요?

june0122 replied:
전자는 특정 포지션 한 곳만의 아이템이 변경되었음을 어댑터에 알려주고, 후자는 인자로 넘겨준 범위에서 변경된 데이터들을 한 번에 알려줍니다. 한 번 데이터를 불러올때 20개의 아이템을 불러오는데 item이 add 될때마다 notify하는 것보다 추가된 20개를 한 번에 notify하는 것이 효율적이라 생각하였습니다.

june0122 committed:
refactor: 코드 패턴 통일성을 위해 `_`가 없는 값도 동일하게 어노테이션 적용

june0122 committed:
style: 리사이클러뷰 아이템의 너비값 수정

june0122 committed:
feat: 페이징 시 로딩 표시를 위한 프로그래스바를 담을 뷰 타입 추가

아이템과 프로그래스바를 위한 뷰타입이 2개 필요한데, 이를 위해선 ViewHolder를 각각의 뷰타입에 맞게
생성해야 한다.

june0122 committed:
feat: 라이브러리를 사용하지 않고 Paging 기능 추가하기

가끔씩 페이지 경계에서 중복된 아이템이 보이는 버그를 수정해야 함.

june0122 committed:
style: 페이징 시 로딩 표시를 위한 프로그래스바가 포함된 리사이클러뷰 아이템 정의

june0122 committed:
fix: 스크롤 시 발생하는 IndexOutOfBoundsException 의심 지점 수정

`notifyItemRemoved(...)`를 하지 않았을 때 프로그래스바가 사라지고 아이템들이
등장하는 애니메이션이 자연스러우므로 사용하고도 유사한 결과가 나올 수 있도록 연구해보기.

june0122 commented:
### Paging 라이브러리 사용하지 않고 직접 구현하기

> 좌측 : 문제 없이 구동되는 영상, 우측 : 페이지 경계에 중복 아이템 발생 버그
<p align = 'center'>
<img width = '250' src = 'https://user-images.githubusercontent.com/39554623/155016310-d3e9f459-e40f-4818-9b0a-8711d6e953bf.gif'>
<img width = '250' src = 'https://user-images.githubusercontent.com/39554623/155016440-a4a91053-65db-4864-9b02-00801214ddd4.gif'>
</p>

- 빠르게 스크롤 시 페이지 경계에서 중복된 아이템이 보이는 버그 처리
- Github API 호출 횟수 초과 시 Toast로 띄워서 보여주기
  - "message":"API rate limit exceeded for 116.XX.XXX.XX.

june0122 committed:
fix: 스크롤 이벤트가 발생하는 도중에 어댑터의 내용을 메인 스레드에서 변경하는 코드 수정

View.post(Runnable)을 사용하여 Runnable을 메시지 큐에 추가하여 UI 스레드에서
실행될 수 있게 처리.

추가적으로 API rate limit 초과했을때 HTTP 403 응답 코드가 발생하면 토스트 메시지를
띄우도록 기능을 추가함.

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/fragment/PlantListFragment.kt:
```diff
@@ -47,6 +67,23 @@ class PlantListFragment : Fragment() {
+                // 마지막 아이템이 완전히 보일 때 해당 포지션을 반환
+                val lastVisibleItemPosition = layoutManager.findLastCompletelyVisibleItemPosition()
+
+                if (!binding.rvPlantList.canScrollVertically(1) && lastVisibleItemPosition == itemCount) {
```
이 코드에서 사용하신 `canScrollVertically`은 최하단:1 or 최상단:-1으로 만 노출됩니다.
 이 경우 사용자가 스크롤 업을 순간적으로 했다가, 바로 다시 최하단으로 갈 경우 1번 더 동일한 데이터를 불리게할 수 있겠죠. 당연히 연속적으로 스크롤업을 빠르게 할 경우도 최하단에서 바로 갱신과 동시에 최하단일 수 있으니 이 역시 동일한 값이 1회 더 불릴 수 있겠고요.
 
 해결하려면 실제 네트워크 비동기 상태가 끝날때를 확이하는 코드가 필요하겠습니다.

Commit:
MVVM 자료

june0122 committed:
feat: ViewModel로부터 받은 메시지를 뷰에 표시하기 위한 Event 클래스 정의

june0122 committed:
refactor: Toast 메시지 확장 함수 정의

june0122 committed:
refactor: 파라미터가 있는 ViewModel 객체 생성을 위해 Factory 클래스 구현

ViewModelProvider.Factory 인터페이스를 구현한 Factory 클래스 정의

june0122 committed:
refactor: 예제 ViewModel 클래스 삭제

june0122 committed:
refactor: PlantListViewModel 패키지 분리

june0122 committed:
refactor: 뷰 컨트롤러에 할당된 과도한 책임을 ViewModel로 분리

june0122 committed:
chore: build.gradle 버전 업데이트

june0122 committed:
refactor: 예제 코드 삭제

june0122 committed:
style: 상세 화면의 background color 추가

june0122 committed:
fix: LiveData에 기존값이 재통보되는 문제를 해결하기 위한 Event Wrapper 추가

june0122 committed:
refactor: 뷰 컨트롤러에 정의된 UI 관련 데이터 분리

june0122 committed:
resource: dimens.xml에 마진값들 추가

june0122 committed:
feat: 로딩 프로세스바에 포커싱될 수 있도록 smoothScroll 기능 추가

june0122 committed:
fix: 프로그래스바가 삭제되면서 좌표값이 이동해 흔들리는 버그 수정

프로그래스바 아이템 삭제 시 position 값이 -1이 되어 프로그래스바가 흔들리는 문제 수정
spanCount가 홀수일 때 여전히 프로그래스바가 흔들리는 문제가 있으므로 수정 필요

june0122 committed:
refactor: 프로그래스바가 데이터 로딩 완료 후에 삭제될 수 있도록 리팩토링

june0122 committed:
fix: 리사이클러뷰 애니메이션을 없애는 것으로 프로그래스바 삭제 시 버그 수정

june0122 committed:
refactor: GithubService의 구현체를 변수를 통해 접근할 수 있도록 변경

동반 객체 내에서 create() 메서드를 통해 생성했던 GithubService의 구현체를
변수를 통해 더욱 간단히 접근할 수 있도록 변경

june0122 committed:
refactor: sealed class를 통해 하위 클래스의 종류를 제한하여 아이템 타입 표시

리사이클러뷰의 아이템 중 프로그래스바가 담긴 아이템을 표현하기 위해 더미 Plant 데이터를
사용하던 코드에서 sealed class인 PlantData를 만들어 이 클래스를 상속하는 Plant,
Progress라는 클래스를 선언. 이로써 이 두 개의 클래스 외에 PlantData 클래스를 상속받는
다른 클래스가 없다는 것을 컴파일러에게 알려준다.

june0122 committed:
refactor: `viewModel.items`를 `items`로 맵핑하지 않고 사용하도록 수정

june0122 committed:
refactor: sealed class `PlantData`를 정의하여 아이템 타입 체크법을 리팩토링

june0122 committed:
refactor: Service와 Retrofit 인스턴스 생성 코드 분리

june0122 committed:
refactor: 상수값을 동반 객체 내에 정의하기

추후 코드의 규모가 커졌을 때 중복 코드 방지도 되고 좋다.

june0122 committed:
refactor: ViewModel 내부에 리턴 타입이 Unit인 함수로만 구성 및 함수 분리

june0122 committed:
Merge pull request #5 from f-lab-edu/mvvm

뷰 컨트롤러에서 ViewModel로 뷰 관련 데이터 분리
